File: src_ex/Main6.ex
defmodule Main6 do

    require Monad.Error

    @spec main :: {:ok, BookShop.Order.t} | {:error, term}
    def main do
        BookShop.test_data |> handle_create_order
    end


    @spec handle_create_order(map) :: {:ok, BookShop.Order.t} | {:error, term}
    def handle_create_order data0 do
        Monad.Error.m do
            data <- BookShop.validate_incoming_data data0
            let %{"cat" => cat0, "address" => address0, "books" => books0} = data
            cat <- BookShop.validate_cat cat0
            address <- BookShop.validate_address address0
            let books1 = Enum.map books0,
                    fn %{"title" => title, "author" => author} ->
                        BookShop.get_book title, author
                    end
            books2 <- Pipeline.sequence books1
            return BookShop.create_order cat, address, books2
        end
    end

end

File: src_erl/main_6.erl
-module(main_6).
-compile({parse_transform, do}).

-export([main/0]).


-spec main() -> {ok, books_shop:order()} | {error, term()}.
main() ->
    handle_create_order(books_shop:test_data()).


-spec handle_create_order(map()) -> {ok, books_shop:order()} | {error, term()}.
handle_create_order(Data0) ->
    do([error_m ||
        Data <- books_shop:validate_incoming_data(Data0),
        #{<<"cat">> := Cat0, <<"address">> := Address0, <<"books">> := Books0} = Data,
        Cat <- books_shop:validate_cat(Cat0),
        Address <- books_shop:validate_address(Address0),
        Books1 = lists:map(
            fun(#{<<"title">> := Title, <<"author">> := Author}) ->
                books_shop:get_book(Title, Author)
            end,
            Books0),
        Books2 <- pipeline:sequence(Books1),
        books_shop:create_order(Cat, Address, Books2)
    ]).

